src/plugins/internalnotification/internalnotification.c

```diff
@@ -0,0 +1,312 @@
+	{
+		KeySet * contract = ksNew (
+			30, keyNew ("system/elektra/modules/internalnotification", KEY_VALUE,
+				    "internalnotification plugin waits for your orders", KEY_END),
+			keyNew ("system/elektra/modules/internalnotification/exports", KEY_END),
+			keyNew ("system/elektra/modules/internalnotification/exports/get", KEY_FUNC, elektraInternalnotificationGet,
+				KEY_END),
+			keyNew ("system/elektra/modules/internalnotification/exports/set", KEY_FUNC, elektraInternalnotificationSet,
+				KEY_END),
+			keyNew ("system/elektra/modules/internalnotification/exports/open", KEY_FUNC, elektraInternalnotificationOpen,
+				KEY_END),
+			keyNew ("system/elektra/modules/internalnotification/exports/close", KEY_FUNC, elektraInternalnotificationClose,
+				KEY_END),
+
+			// Export registerInt function and required plugin handle
+			keyNew ("system/elektra/modules/internalnotification/exports/elektraInternalnotificationRegisterInt", KEY_FUNC,
+				elektraInternalnotificationRegisterInt, KEY_END),
+			keyNew ("system/elektra/modules/internalnotification/exports/handle", KEY_BINARY, KEY_SIZE, sizeof handle,
+				KEY_VALUE, &handle, KEY_END),
+
+#include ELEKTRA_README (internalnotification)
+
+			keyNew ("system/elektra/modules/internalnotification/infos/version", KEY_VALUE, PLUGINVERSION, KEY_END), KS_END);
+		ksAppend (returned, contract);
+		ksDel (contract);
+
+		return 1;
+	}
+
+
+	PluginState * pluginState = elektraPluginGetData (handle);
+	ELEKTRA_ASSERT (pluginState != NULL, "plugin state was not initialized properly");
+
+	elektraInternalnotificationUpdateRegisteredKeys (pluginState, returned);
+
+	return 1;
+}
+
+/**
+ * Updates registrations with data written by the application.
+ * Part of elektra plugin contract.
+ *
+ * @param  handle    plugin handle
+ * @param  returned  key set containing current data from the application
+ * @param  parentKey key for errors
+
+ * @retval 1 on success
+ * @retval -1 on failure
+ */
+int elektraInternalnotificationSet (Plugin * handle, KeySet * returned, Key * parentKey)
+{
+	ELEKTRA_LOG_DEBUG ("keyName=%s", keyName (parentKey));
+
+	PluginState * pluginState = elektraPluginGetData (handle);
+	ELEKTRA_ASSERT (pluginState != NULL, "plugin state was not initialized properly");
+
+	elektraInternalnotificationUpdateRegisteredKeys (pluginState, returned);
+
+	return 1;
+}
+
+/**
+ * Initialize data plugin data structures.
+ * Part of elektra plugin contract.
+ *
+ * @param  handle         plugin handle
+ * @param  parentKey      key for errors
+ *
+ * @retval 1 on success
+ * @retval -1 on failure
+ */
+int elektraInternalnotificationOpen (Plugin * handle, Key * parentKey ELEKTRA_UNUSED)
+{
+	PluginState * pluginState = elektraPluginGetData (handle);
+	if (pluginState == NULL)
+	{
+		pluginState = elektraMalloc (sizeof *pluginState);
+		if (pluginState == NULL)
+		{
+#if PLUGIN_INTERNALNOTIFICATION_VERBOSE
+			fprintf (stderr, "elektraInternalnotificationOpen elektraMalloc failed!\n");
```

Comment:
please use logging also here and you should set an error for outside world. (Which does the logging already for you)